### DISCUSSION TITLE
Any way for set_rhs to query a black-box function?

### DISCUSSION THREAD

**Discussion by shushmanLacuna** [do-mpc/do-mpc#207] [Q&A] [answered]
Thanks so much for this useful library!

All of the examples I've seen require the dynamics model to be an explicit algebraic equation in terms of the state and control, e.g., https://github.com/do-mpc/do-mpc/blob/master/examples/oscillating_masses_discrete/template_model.py#L66

Is there (or will there be) any support for models that can query some black-box function `foo(x,u)` to get the next state? For instance, a learned statistical or neural network model that encodes the implicit system dynamics of some unknown system?

**Comment by 4flixt** [comment, marked as answer]
Hi @shushmanLacuna ,

thank you for using do-mpc! 
There is no way to use a generic black-box function as the system model in do-mpc. The problem is that for efficient optimization the tool must be able to compute gradients with Automatic Differentiation as provided by CasADi.

However, we are using neural networks as system models in our research. While you think of them as "black box" they have a clear set of instructions (simple matrix vector multiplications and nonlinear activation functions) that you can implement yourself in do-mpc.
The trick is to train a NN (or any other ML model) and then retrieve the parameters for re-implementation. 

It is on our roadmap to have an automatic conversion from e.g. TF/Keras to do-mpc.

Best
4flixt

**Reply by jrbull** [reply]
I have a similar query. My company has a well-developed in-house numerical simulation library in Python. I don't want to have to rewrite it as symbolic DAE models. The functions are not very mathematically complex, and I have been successfully using IPOPT to optimise the models with approximate (numerical) Jacobians. Is it a total no-go to import numerical RHS functions to do-mpc?
